analysis/token/synonym/synonym.go

```diff
@@ -0,0 +1,361 @@
+// the FST, validPaths is appended with the new path that was found.
+// Fuzziness is the maximum Levenshtein distance allowed between matchTry and the word fuzzily matched in the FST.
+//
+// Basic logic followed is as follows:
+//   - At the state in the FST where the character in the input string fails to match,
+//     we start a DFS searching for states that end a word
+//   - which are states that either do not have any outgoing transition or
+//     have a transition for a space
+//   - Since it is a FST, multiple such paths will be found, and each path's word is checked for
+//     edit distance criteria satisfaction before it is added to validPaths
+func fuzzyMatch(path *fstPath, matchTry string, validPaths []*fstPath, fst *vellum.FST, fuzziness int) ([]*fstPath, error) {
+
+	// stack for iterative DFS
+	pathStack := []*fstPath{path}
+	var pathIsValid bool
+	for len(pathStack) > 0 {
+		path = pathStack[len(pathStack)-1]
+		pathStack = pathStack[:len(pathStack)-1]
+		pathIsValid = false
+		// if input FST path ends with a word and if the word fuzzily matches the matchTry, append the validPaths with the input FST path.
+		endsWithWord, err := pathEndsWithWord(path, fst)
+		if err != nil {
+			return nil, err
+		}
+		if endsWithWord {
+			if search.LevenshteinDistance(string(path.word), matchTry) <= fuzziness {
+				validPaths = append(validPaths, path)
+				pathIsValid = true
+			}
+		}
+		// perform a DFS from the current state of the input FST path.
+		// even if the path ends with a word, the DFS is performed since there there may be
+		// another transition apart from a space at the state in the path.
+		//
+		// this condition is a shortcut to discard certain fst paths where it becomes known that the word
+		// will have a edit distance > fuzziness to matchTry, as path does not end with a word and
+		// path.word is already exceeding len(matchTry) by 'fuzziness' amount and any further
+		// exploration along this path will need removal of characters >'fuzziness'
+		if len(path.word)-len(matchTry) <= fuzziness {
+			transitions, err := fst.GetTransitionsForState(path.state)
+			if err != nil {
+				return nil, err
+			}
+			// for each transition from the current state of the input FST path, perform a DFS.
+			// if the transition is not a space, get a new  fstPath from the pool, and insert it
+			// into the stack with the updated parameters
+			for _, character := range transitions {
+				if character != SeparatingCharacter {
+					newState, newOutput := fst.AcceptWithVal(path.state, character)
+					newPath := pathPool.Get().(*fstPath)
+					newPath.state, newPath.output, newPath.word = newState, path.output+newOutput, append(path.word, character)
+					pathStack = append(pathStack, newPath)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+		}
+		// only discard fstPath if it was not added to validPaths
+		if !pathIsValid {
+			pathPool.Put(path)
```

Comment:
before putting the path back into the sync pool, its always recommended to reset the values in it.

Reply:
Also please check the case where you'd be putting the path back into the pool after processing a valid path. If you don't do that, the pool would be empty across multiple queries and the reusability wouldn't be effective, because now that path would return to the heap rather than the sync pool.

Reply:
i didnt get you bro
like once i get a set of valid paths and fuzzy search is complete ,
i must iterate over all valid paths, reset it, and put it back in the pool before exiting the main Filter function?

Reply:
either that or better yet every time a path from the validPath results in a match (I see you're incrementing the pathIndex in that case) you could put it back in the pool since it's no longer going to be used because you're going to process the next path in the validPaths slice.

Reply:
gg ez done 
gomu gomu no code